teps/0050-ignore-task-failures.md

```diff
@@ -187,6 +194,222 @@ control over the task definitions but may desire to ignore a failure and continu
+
+```
+"Tasks Completed: A (Failed: B (C is ignored), Cancelled D), Skipped: E"
+```
+
+Example Input:
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata: 
+  name: demo-pipeline-run
+spec:
+  pipelineSpec:
+    tasks:
+    - name: task1
+      OnFail: continue
+      taskSpec:
+        steps:
+          - image: alpine
+            name: exit-with-1
+            script: |
+              exit 1
+    - name: task2
+      taskSpec:
+        steps:
+          - image: alpine
+            name: exit-with-0
+            script: |
+              exit 0
+```
+
+Example Output:
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+...
+status:
+  completionTime: "2022-08-15T17:26:15Z"
+  conditions:
+    - lastTransitionTime: "2022-08-15T17:26:15Z"
+      message: "Tasks Completed: 2 (Failed: 1 (1 is ignored), Cancelled 0), Skipped: 0"  
+      reason: Succeeded
+      status: "True"  # The failed task is considered "successful" when determining the state of pipelineRun
+      type: Succeeded
+  pipelineSpec:
+    ...
+  taskRuns:
+    demo-pipeline-run-task1:
+      pipelineTaskName: task1
+      status:
+        completionTime: "2022-08-15T17:26:13Z"
+        conditions:
+          - lastTransitionTime: "2022-08-15T17:26:13Z"
+            message: [Task is failed but ignored] ...
+            reason: Failed
+            status: "False" # The task is failed even OnFail is set to continue
+            type: Succeeded
+          ...
+    demo-pipeline-run-task2:
+      pipelineTaskName: task2
+      status:
+        completionTime: "2022-08-15T17:26:15Z"
+        conditions:
+          - lastTransitionTime: "2022-08-15T17:26:15Z"
+            message: All Steps have completed executing
+            reason: Succeeded
+            status: "True"
+            type: Succeeded
+          ...
+
+```
+
+### Ignored Failed Tasks with Retry
+Setting ```Retry``` and ```OnFail``` to ```continue``` at the same time is not allowed in this iteration of the TEP, as there is no point to retry a task that allows to fail. Pipeline validation will be added accordingly. We can support retries with ignored failed task in the future if needed.
+
+### Tasks with Missing Resource Dependency
+In the following example, the first task fails to produce a result (due to step error [details](https://github.com/tektoncd/pipeline/issues/3749)) that is going to be consumed by the second task. The second task will be skipped with reason ```Results were missing``` irrelevant of its OnFail type. This behaviour is consistent with [Guarding a Task only](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#guarding-a-task-only). When we add support for [default Results](https://github.com/tektoncd/community/blob/main/teps/0048-task-results-without-results.md), then the resource-dependent Tasks may be executed if the default Results from the skipped parent Task are specified. 
```

Comment:
This paragraph is little confusing and inconsistent with `step.onError`.

The linked issue has been partially fixed in which pipeline results are emitted even when a task fails. Pipeline results are initialized with the task results from all successful tasks until a task fails.

`step.onError` does support emitting results and consuming them - https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#produce-a-task-result-with-onerror

As long as a task has initialized results before failing, they can be made available for consumption. This only should be supported with `task.onFail: continue`.

When a pipeline author has chosen a strategy where a task failure is expected, the author must have complete understanding of what the `result` could mean and how to consume that `result` if it's initialized before failing.

For example, `git-clone` produces `commit` as a task result. If pipeline author chooses to ignore failure from `git-clone`, the author is aware that the `commit` **might/might not** be initialized depending on the failure. Now, as a pipeline author, when a `git-clone` failed for some reason but `commit` is available, its guaranteed that my repo was cloned as expected. If the `commit` is not initialized and `git-clone` failed, the repo was not cloned for some reason.

Another example is [boskos-acquire](https://github.com/tektoncd/catalog/blob/main/task/boskos-acquire/0.1/boskos-acquire.yaml). As a pipeline author, I can choose to ignore failure from `boskos-acquire`. If `boskos-acquire` failed for some reason to create heartbeat pod but was able to lease resources, `leased-resource` will be initialized and can be made available to the `cleanup` task if its initialized.

Reply:
> As long as a task has initialized results before failing, they can be made available for the consumption. This only should be supported with `task.onFail: continue`.
> 
> When a pipeline author has chosen a strategy where a task failure is expected, the author must have complete understanding of what the `result` could mean and how to consume that `result` if it's initialized before failing.

Thanks for your input, @pritidesai! This is a good point and the question is essentially: **Should we emit results for failed tasks with onFail:continue**? And this may need some discussion.

Currently we don't emit results for failed tasks. There are some [discussions](https://github.com/tektoncd/pipeline/issues/3749) around it and I'm not sure why we ended up not implement it.  Looking at the requirement 3 of this TEP: ```the task would be considered "successful" for the purposes of determining the status of the pipeline```, it seems to me we should respect the existing behaviour about emitting result, and change behaviour of emitting task is another discussion/issue/TEP? Is it a good idea that we respect the existing behaviour as default, and create another TEP/Issue specifically deal with the emitting result behaviour if we can such request from users?

In terms of consistency, I think ```step.OnError``` and ```task.OnFail``` are different things. ```step.onError``` emits the result only when the overall task run status is success, but in ```task.OnFail``` we keep the task run status as failed even when it is set to  ```continue```

Thoughts? @jerop @vdemeester ?

Reply:
I think we have an opportunity here to design this right for many different use cases we have seen. I would recommend attempting to solve here rather than delegating it to a separate TEPs.

 `task.onError` is a failure strategy and no different from `step.onError`. Pipeline is not permitted to continue running after a first failure is encountered. With `task.onError`, even though the task is failed (container exit code stays non-zero), the pipeline is allowed to continue executing the rest of the tasks.

Reply:
@pritidesai @lbernick @jerop 

I have added a new section [Emit Results from Ignored Failed Tasks](https://github.com/QuanZhang-William/community/blob/tep-0050-add-proposed-design/teps/0050-ignore-task-failures.md#emit-results-from-ignored-failed-tasks) and updated [Task with Missing Resource Dependency](https://github.com/QuanZhang-William/community/blob/tep-0050-add-proposed-design/teps/0050-ignore-task-failures.md#tasks-with-missing-resource-dependency) section based on our discussion. Please take a look